### PR TITLE
foundation for multi-output sources

### PR DESCRIPTION
**Pull request by petrosagg**
This PR implements the foundations for supporting sources that can output multiple streams that are written directly to multiple storage collections. In order to do that almost all the parts of materialize needed to be touched.

In order to follow along during review it's helpful to spell out how multi-output sources are modelled.

### Adapter

The main departure from how things currently are is breaking the 1-1 relationship between source objects in the catalog and running ingestions. In particular, this is the diff of the catalog source definition:

```diff
 pub struct Source {
-    pub source_desc: SourceDesc,
+    pub ingestion: Option<Ingestion>,
    ..
 }
```

So now materialize has sources with an associated ingestion and sources without. Sources without ingestion are referred to in the code as "subsources" and only exist because some other source with an associated ingestion is writing data into them.

The relationship between sources and subsources is recorded by the following sequence of SQL commands, which all happen in a DDL transaction:

```sql
CREATE SUBSOURCE "table1" (a int NOT NULL);
CREATE SUBSOURCE "table2_alias" (b text);
CREATE SOURCE "my_source" FROM ... FOR TABLES ("table1" INTO "table1", "table2" INTO "table2_alias");
```

The above sequence of commands is created automatically as part of purification when users type a single `CREATE SOURCE .. ` statement. Therefore from the user's perspective a single `CREATE SOURCE` statement will appear to immediately and atomically create a group of related objects, if the statement included `FOR TABLE ..` or `FOR ALL TABLES`.

### Storage

Ingestions have also changed to be able to describe an arbitary number of outputs streams of which a subset (or all of them) are being written into storage collections. In particular, this is the diff of the ingestion description:

```diff
 pub struct IngestionDescription<S = ()> {
     /// The source description
     pub desc: SourceDesc
     /// Additional storage controller metadata needed to ingest this source
-    pub storage_metadata: S,
-    /// The relation type this ingestion should produce
-    pub typ: RelationType,
+    pub ingestion_metadata: S,
+    /// Collections to be exported by this ingestion. Each GlobalId is associated with the ordinal
+    /// of the output stream it should be connected to
+    pub source_exports: BTreeMap<GlobalId, (usize, S)>,
 }
```

Conceptually, a `SourceDesc` describes the dataflow that connects to the external system and brings in a number of output streams. Then, each `source_export` is a storage collection that receives the output stream it is associated with.

## Review aid

Here I describe the code changes in more detail to aid reviewing.

### Subsource SQL syntax

Starting with the easy parts the following additions to SQL syntax have been implemented:

```sql
CREATE SUBSOURCE "foo" (a int NOT NULL);
```

A subsource is just a source that doesn't have an associated ingestion, just an expected schema and constraints. Its definition is very similar to tables but they cannot be inserted to by the users. In fact, they are planned identically to sources (i.e there is no
`Plan::CreateSubsourcePlan`). As far as the storage controller goes this looks identical to a table that is never appended to.

Subsource statements only exist so that we can generate them during purification and store them in the catalog.

### Sub-table selection syntax

```sql
CREATE SOURCE ... FOR ALL TABLES;
-- or
CREATE SOURCE ... FOR TABLES ("foo" AS "bar");
```

All create source statements can be followed by a sub-table selection clause that lets user select which of the outputs they are interested in, if the source is a multi-output source.

Using this syntax with sources that don't provide additional output streams is an error, and so is *not* providing it for sources that do. In other words, if the source has multiple outputs you *have* to specify the sub-tables you want.

The sub-table selection syntax has the two variants illustrated above that are meant to be used by users, and a third variant that is only ever constructed by purification. That third variant looks like this:

```sql
CREATE SOURCE ... FOR TABLES ("foo" INTO "bar");
```

Their only difference is in the AST definition where the `AS` version treats the alias target as a raw identifier whereas the `INTO` version treats the alias target as a fully resolved catalog object.

The reason we need the `AS` and `INTO` variants is that when the user types their `CREATE SOURCE` statement the subsource catalog objects don't exist yet, and therefore we must parse the identifers as raw identifiers. On the other hand, once we have purified the statement and created the subsources we want the `CREATE SOURCE` statement to truly refer to these newly formed catalog objects so that dependency tracking and catalog bootstrapping works correctly.

Therefore as part of processing a `CREATE SOURCE` statement with table sub-selection we first create the subsources and then re-write the alias section to be the `INTO` variant before writing it down in the catalog.

### Mini DDL transactions

As mentioned above we want the user experience of running a `CREATE SOURCE` statement with a multi-output source to be a group of objects atomically appearing. Materialize doesn't yet support DDL transactions so a shortcut has been taken. It works as follows:

Purification receives a `CreateSourceStatement<Aug>` to purify. Purification is when we reach out to external systems to determine what data there is available. For postgres sources this is when we learn the list of available upstream tables. What we want purification to do is to return a list of `CreateSubsourceStatement<Aug>`, one for each upstream table that we want to ingest, and alter the original `CreateSourceStatement<Aug>` to have a `FOR TABLES ("foo" INTO "bar")` suffix.

This last step is impossible with some trick. `Aug` type statements store the final `GlobalId` of all objects they refer to, but we want to refer to subsource objects that have not been created yet.

The trick is that purification mints `GlobalId::Transient` identifiers for all the subsources it requests to build and then produces the `INTO` syntax by pointing it to those ids. Concretely, this is the simplified function signature of `purify_create_source`:

```rust
fn (CreateSourceStatement<Aug>) -> (Vec<(GlobalId, CreateSubsourceStatement<Aug>)>, CreateSourceStatement<Aug>)
```

The `CREATE SOURCE` statement with the transient id references cannot be ran as-is so here is the final step of the trick:

When the coordinator receives the list of subsources, it mints a new user `GlobalId`s and stores the decition in a map of `transient_id -> global_id`. Then an AST folder parameterized with this id allocation is ran on the AST of the `CREATE SOURCE` statement that replaces all the transient global id references with the concrete user global ids the coordinator has selected.

Finally, with correctly formed AST statements, all of them are planned and then sequenced as a single catalog transaction.

#### Multi-output SourceReader

The core change in storage is the type of `SourceMessage`, which is the message type all source readers are tasked to produce.

```diff
 pub struct SourceMessage<Key, Value, Diff> {
+    /// The output stream this message belongs to. Later in the pipeline the stream is partitioned
+    /// based on this value and is fed to the appropriate source exports
+    pub output: usize,
     ...,
 }
```

This added field allows a source reader to indicate for which of its output streams this message is for. When the ingestion dataflow is rendered we first reclock the `SourceMessages` without special treatment of the `output` field and then, using timely's partition operator, we create as many timely streams as there are outputs.

```diff
+    let (ok_muxed_stream, err_stream) =
+        reclocked_stream.map_fallible("reclock-demux", |(output, r)| match r {
+            Ok(ok) => Ok((output, ok)),
+            Err(err) => Err(err),
+        });
+
+    let ok_streams = ok_muxed_stream.partition(u64::cast_from(num_outputs), |(output, data)| {
+        (u64::cast_from(output), data)
+    });
```

Once the output streams have been partitioned they receive the same treatment. I.e for each stream the same decoding and envelope processing steps will be rendered and instead of a single output `render_source` provides a vector of output streams.

Finally, a persist sink is rendenred for each of the `source_exports` declared in the ingestion description that is connected to its associated output index:

```diff
+            for (target, (output_idx, metadata)) in description.source_exports {
+                let (ok, err) = &outputs[output_idx];
+                let source_data = ok.map(Ok).concat(&err.map(Err));
+                crate::render::persist_sink::render(
+                    region,
+                    target,
+                    metadata,
+                    source_data,
+                    storage_state,
+                    Rc::clone(&token),
+                );
+            }
```

#### Resumption frontier calculation

Calculating the resumption frontier of an ingestion has also changed slightly. We now need to consider the upper frontier of all output shards and the upper frontier of all state shards. The logic for calculating it has been moved on the `IngestionDescription` struct that has all the required information to compute it.

Signed-off-by: Petros Angelatos <petrosagg@gmail.com>

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  -

### CODE REVIEW

**Comment by philip-stoev** [comment]
@petrosagg please advise once this becomes ready for testing, as it appears to be a major refactor. I will also need to add/modify our existing tests that touch Postgres. Thank you!

**Comment by petrosagg** [comment]
@philip-stoev will do. This PR particular leaves all sources as single-output so the current test suite should verify that I didn't break anything. I'll ping you in the next one that changes pg sources from single-output to multi-output which will further exercise all this machinery

File: src/storage/src/types/sources.rs
Comment (guswynn) [reply src/storage/src/types/sources.rs:1016]: Also, is moving the source arity into the env desc just an additional refactor, or is their some fundamental reason it has to happen in this pr?

File: src/storage/src/types/sources.rs
Comment (guswynn) [inline src/storage/src/types/sources.rs:973]: might want a comment here explaining that this case is only hit by the `None` case (or better yet, move them into different branches

**Comment by sploiselle** [comment]
@petrosagg You mention that we go through a lot of pain to support creating sources + subsources atomically, but this might be more easily supported in the adapter. Did you see how `adapter::coord::sequencer::sequence_create_views` handles generating multiple views at once? If so, would love to understand why that approach is incommensurate with the source + subsource relationship.

**Review by guswynn — APPROVED** [review]
Overall the storage changes look reasonable to me; the dataflow changes are fairly straighforward; I had some questions and some comments (that can mostly be fixed later), but one has to be answered before merging:
- are we relying on the final env wipe to merge this? there are incompatible changes to some proto and sql schemas (the proto ones are less important, as they are just over the wire)



As for the sql changes, your description seems reasonable to me, and I looked at the code changes, and thing struck me as problematic, other than ensuring users themselves can't type `CREATE SUBSOURCE`, but someone else with more adapter/sql expertise will likely need to sign off, particularly on the changes in `src/sql/src/plan/statement/ddl.rs`

**Comment by petrosagg** [comment]
@sploiselle Yes, I've seen the create views machinery (I wrote it :blush:). The difference is that when a user runs `CREATE VIEWS FROM SOURCE "foo"` the source the views depend on has already been created (by some previous `CREATE SOURCE "foo"` statement). This means that the reference to `"foo"` in there can be resolved by looking it up in the catalog and everything works out.

In the case where we want to create sources and subsources in one step we actually want to introduce a catalog dependency between the source and the subsources but none of them has been created yet so `mz_sql::names::resolve` can't yet resolve the names to objects and the planner can't look up the subsources in the catalog to inspect them. This leads the slightly trickly implementation that is in this PR.

**Comment by sploiselle** [comment]
@petrosagg If all this needs to do is insert a dependency for the subsources, could that be done in the adapter? Like we could determine the dependencies as normal, and then for the subsources just append an additional dependency on the source itself [here](https://github.com/MaterializeInc/materialize/pull/15159/files#diff-baf2f251b60cc61cc509f17df66975565207909931d35f7c20d34b740f9ef3c0R460). Believe that you've thought of this but just trying to figure out where the challenge lies.

**Comment by petrosagg** [comment]
@sploiselle it's a good question! The problem lies within the fact that the `depends_on` of a catalog item is derived by parsing its `create_sql` and resolving the object names it refers to. In other words it's the AST itself of the create source statement that contains references to the subsources and `depends_on` is just a cache so that we don't traverse the AST when we want to see its dependencies. This means that I can't change the `depends_on` directly, but the data it is derived from.

The `CreateSourceSubsource` is defined like this:

```rust
pub enum CreateSourceSubsource<T: AstInfo> {
    /// An unaliased subbsource
    Bare(UnresolvedObjectName),
    /// An subbsource aliased to a different name
    Aliased(UnresolvedObjectName, UnresolvedObjectName),
    /// An subbsource fully resolved to the target catalog object it will be written to
    Targeted(UnresolvedObjectName, T::ObjectName),
}
```

And after purification all subsources must be described by the third variant (`Targeted`) which requires a valid `<Aug as AstInfo>::ObjectName`, which requires a `GlobalId` before we even hit the planner.

This is why during purification I assign transient ids to all the subsources and create the `CreateSourceSubsource::Targeted` references in the source AST using those transient ids. Then the coordinator allocates real ids for all the transient ones, traverses the AST to replace them wherever they might be used, and only then invokes the planner. 

The correct `depends_on` then just follows by having a correctly formed AST pointing to the right objects.

**Comment by sploiselle** [comment]
@petrosagg While that's true in the general case, there's nothing to stop you from plumbing all of this down into the adapter and further developing the `depends_on` there. This is vaguely how clusters with initially instantiated replicas work.

My contention here, maybe more explicitly, is that you don't need `CreateSubsourceStatements` whatsoever; the schemas and identifiers of the subsources could be bundled into the `CreateSourceStatement` and expanded in `sequence_create_source`, with the newly-expanded subsources' sole dependency being the "main" source. We don't allow foreign key constraints, etc., on subsources, so we know there aren't any other dependencies. If I've got this dependency inverted (i.e. the dependency goes from source -> subsource), that could still be pushed into the source's `depends_on`.

It seems like this code introduces a lot of complexity to maintain the invariant that the `depends_on` is generated from the AST, but I don't understand why that is a hard requirement.

File: src/adapter/src/coord/dataflows.rs
Comment (petrosagg) [reply src/adapter/src/coord/dataflows.rs:418]: Yeah, I'll add a todo because to do that the catalog needs to maintain an inverse mapping of subsources -> sources so that we can look it up.

**Comment by sploiselle** [comment]
Ah, ok--when we serialize the item in the catalog, it needs to have all of its `GlobalId` references intact. This is more naturally done during name resolution, rather than grafting the references in place in the adapter; the latter would have been my preference rather than the kind of torture this endures. But, this has the benefit of retaining the clear separation of concerns, so I see the value in this approach and am satisfied on that front.

File: src/storage/src/render/sources.rs
Comment (petrosagg) [reply src/storage/src/render/sources.rs:194]: changed

File: src/storage/src/render/sources.rs
Comment (petrosagg) [reply src/storage/src/render/sources.rs:95]: Changed this to just call a `SourceDesc::num_outputs()` function with the comment moved in there

File: src/storage/src/types/sources.proto
Comment (petrosagg) [reply src/storage/src/types/sources.proto:232]: yes

File: src/storage/src/types/sources.rs
Comment (petrosagg) [reply src/storage/src/types/sources.rs:60]: done

File: src/storage/src/types/sources.rs
Comment (petrosagg) [reply src/storage/src/types/sources.rs:1335]: for this PR yes

File: src/storage/src/types/sources.rs
Comment (petrosagg) [reply src/storage/src/types/sources.rs:1016]: It was necessary but I've lost track of what needed that

File: src/storage/src/types/sources.rs
Comment (petrosagg) [reply src/storage/src/types/sources.rs:206]: fixed

File: src/storage/src/source/source_reader_pipeline.rs
Comment (petrosagg) [reply src/storage/src/source/source_reader_pipeline.rs:1152]: changed

File: src/adapter/src/coord/command_handler.rs
Comment (guswynn) [inline src/adapter/src/coord/command_handler.rs:434]: nice

File: src/storage/src/render/sources.rs
Comment (guswynn) [inline src/storage/src/render/sources.rs:108]: im okay with this todo for now

**Review by guswynn — COMMENTED** [review]
updates look good

**Review by sploiselle — COMMENTED** [review]
I did not have a chance to look at the storage stuff yet

File: src/sql/src/plan/statement/ddl.rs
Comment (petrosagg) [reply src/sql/src/plan/statement/ddl.rs:1392]: The plan is to completely `CREATE VIEWS` in favour of multioutput sources. The only other source using it is the load generator which if I have time I'll migrate too and then delete the `CREATE VIEWS` syntax.

File: src/adapter/src/coord/dataflows.rs
Comment (petrosagg) [reply src/adapter/src/coord/dataflows.rs:416]: It's unused right now but it's because once we find a way to store the id of the parent source we'd like to use `self.catalog` to look it up and find its monotonicity

File: src/adapter/src/coord/sequencer.rs
Comment (petrosagg) [reply src/adapter/src/coord/sequencer.rs:491]: This empty tuple is pervasive when interacting with storage, this is just one of the many places it appears. I agree it is confusing though, I created an issue to address the issue as a whole https://github.com/MaterializeInc/materialize/issues/15199

File: src/sql-parser/src/ast/defs/statement.rs
Comment (petrosagg) [reply src/sql-parser/src/ast/defs/statement.rs:576]: Yeah I think we'd have to create a system table for that because the subsources are created before the source that is writting to them and so they can't have a `FROM SOURCE` reference, the AST name resolution wouldn't work. I'd like to leave that to a future PR if that's ok

File: src/sql/src/plan/statement/ddl.rs
Comment (petrosagg) [reply src/sql/src/plan/statement/ddl.rs]: We can't use `soft_assert` here (or in the other cases) because soft assert doesn't unconditionally assert, or panic, which means that the code path having the `soft_assert` must return a correctly typed value and have the rest of the program continue.

In this case we can't really continue planning so the two options are returning from the planner with an error, or unconditionally panic. I think we don't want to unconditionally panic in this case

File: src/adapter/src/coord/dataflows.rs
Comment (sploiselle) [reply src/adapter/src/coord/dataflows.rs:416]: ty

File: src/adapter/src/coord/sequencer.rs
Comment (sploiselle) [reply src/adapter/src/coord/sequencer.rs:489]: TIL

File: src/sql/src/plan/statement/ddl.rs
Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs]: Ah, sorry, I meant in addition to erroring, i.e. panic if it's reasonable to. Sounds like you'd rather not, so no need.

File: src/sql/src/plan/statement/ddl.rs
Comment (sploiselle) [reply src/sql/src/plan/statement/ddl.rs:1371]: ty for the pointer

File: src/sql-parser/src/ast/defs/statement.rs
Comment (sploiselle) [reply src/sql-parser/src/ast/defs/statement.rs:570]: System table sgtm

**Comment by sploiselle** [comment]
SQL stuff LGTM modulo outstanding design discussions

**Comment by petrosagg** [comment]
Talked with @sploiselle on Slack and since this PR doesn't yet expose the new syntax to users because no source offers multiple outputs I'm going to go ahead and merge it and get the postgres PR ready!

**Pull request merged**
